bluemira/equilibria/fem_fixed_boundary/fem_magnetostatic_2D.py

```diff
@@ -183,3 +200,255 @@ def calculate_b(self):
+        return dolfin.assemble(self.g * dx())
+
+    def _update_curr(self, curr_target):
+        self.k = 1
+        self.k = curr_target / self._calculate_curr_tot()
+
+    def _plot_current_iteration(self, points, i_iter):
+        curr_data = np.array([self.g_func(p) for p in points])
+        self._plot_array(
+            points, curr_data, f"J current at iteration {i_iter}", contour=False
+        )
+
+    def _plot_array(self, points, array, title, contour=True):
+        plot_scalar_field(
+            points[:, 0],
+            points[:, 1],
+            array,
+            levels=20,
+            axis=None,
+            tofill=True,
+            contour=contour,
+        )
+        plt.title(title)
+        plt.show()
+
+    def solve(
+        self,
+        pprime,
+        ffprime,
+        curr_target,
+        dirichlet_bc_function=None,
+        dirichlet_marker=None,
+        neumann_bc_function=None,
+        iter_err_max=1e-5,
+        max_iter=10,
+        relaxation=0.0,
+        verbose=False,
+        plot=False,
+    ):
+        """
+        Solve the G-S problem given pprime and ffprime.
+
+        Parameters
+        ----------
+        pprime: Union[callable, float]
+            pprime as function of psi_norm (1-D function)
+        ffprime: Union[callable, float]
+            ffprime as function of psi_norm (1-D function)
+        curr_target: float
+            Target total plasma current
+        dirichlet_bc_function : Optional[Union[dolfin.Expression, dolfin.Function]]
+            Dirichlet boundary condition function. Defaults to a Dirichlet boundary
+            condition of 0 on the plasma boundary.
+        neumann_bc_function : Optional[Union[dolfin.Expression, dolfin.Function]]
+            Neumann boundary condition function. Defaults to a Neumann boundary
+            condition of 0 on the plasma boundary.
+        dirichlet_marker : int
+            Identification number for the dirichlet boundary
+        iter_err_max: float
+            Convergence criterion value
+        max_iter: inter
+            Maximum number of iterations
+        relaxation: float
+            Relaxation factor for the Picard iteration procedure
+        verbose: bool
+            Whether or not to print
+        plot: bool
+            Whether or not to plot
+
+        Returns
+        -------
+        psi: dolfin.Function
+            dolfin.Function for psi
+        """
+        points = self.mesh.coordinates()
+
+        if plot:
+            self.g_func = self._create_g_func(pprime, ffprime, curr_target)
+            self._plot_current_iteration(points, 0)
+
+        self.g = self._create_g(pprime, ffprime, curr_target)
+
+        super().solve(
+            self.g, dirichlet_bc_function, dirichlet_marker, neumann_bc_function
+        )
+        self._update_curr(curr_target)
+
+        eps = 1.0
+        i = 0  # iteration counter
+        while eps > iter_err_max and i < max_iter:
```

Comment:
As above:



In future, we could use a ConvergenceCriterion (as in equilibria/solve.py), but it is not necessary for this PR.

In future, we could use a more generic version of the PicardIterator (as in equilibria/solve.py) but right now this functionality only makes sense for free boundary equilibria with coils (as they are used to stabilise the problem, in crude terms) - none of which is necessary for a fixed boundary problem, or for this PR.